File: backend/app/routers/agent.py
import json

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from app.core.logging import get_logger
from app.models.schemas import AgentResponse, ChatRequest
from app.services.agent import shopping_agent
from app.services.otp import extract_otp, submit_otp

router = APIRouter(tags=["agent"])
logger = get_logger(__name__)


def _sse(event: str, data: dict) -> str:
    """Format one Server-Sent Event frame."""
    return f"event: {event}\ndata: {json.dumps(data)}\n\n"


@router.post("/agent/chat", response_model=AgentResponse)
async def agent_chat(request: ChatRequest) -> AgentResponse:
    # Deterministic "otp <number>" trigger — handled before the agent so the model
    # never sees or invents OTP codes (mirrors the legacy /chat behaviour).
    latest = request.messages[-1]
    if latest.role == "user":
        otp = extract_otp(latest.content)
        if otp:
            try:
                await submit_otp(otp)
            except Exception as exc:
                logger.exception("OTP submission error")
                raise HTTPException(
                    status_code=502,
                    detail="Could not submit the OTP right now. Please try again.",
                ) from exc
            return AgentResponse(reply=f"Submitted OTP **{otp}**.", thread_id=request.thread_id)

    try:
        return await shopping_agent.run(request.messages, thread_id=request.thread_id)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except Exception as exc:
        logger.exception("Agent endpoint error")
        raise HTTPException(
            status_code=502, detail="The AI service is currently unavailable. Please try again."
        ) from exc


@router.post("/agent/chat/stream")
async def agent_chat_stream(request: ChatRequest) -> StreamingResponse:
    """Same as /agent/chat but streams progress as Server-Sent Events: `status`
    (a tool is running), `results`/`pending_live` (as a search lands), `reply`
    (final text), `done` (the full AgentResponse), or `error`. A streaming 200
    can't change its status code mid-flight, so failures arrive as an `error`
    event rather than an HTTP error."""
    latest = request.messages[-1]

    # OTP short-circuit — mirrors the non-streaming endpoint (model never sees it).
    if latest.role == "user" and (otp := extract_otp(latest.content)):

        async def otp_stream():
            try:
                await submit_otp(otp)
                resp = AgentResponse(reply=f"Submitted OTP **{otp}**.", thread_id=request.thread_id)
                yield _sse("done", resp.model_dump())
            except Exception:
                logger.exception("OTP submission error")
                yield _sse(
                    "error",
                    {"detail": "Could not submit the OTP right now. Please try again."},
                )

        return StreamingResponse(otp_stream(), media_type="text/event-stream")

    async def event_stream():
        try:
            async for event, data in shopping_agent.run_stream(
                request.messages, thread_id=request.thread_id
            ):
                yield _sse(event, data)
        except Exception:
            logger.exception("Agent stream error")
            yield _sse(
                "error",
                {"detail": "The AI service is currently unavailable. Please try again."},
            )

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: worker/src/routers/agent.ts
import { Hono } from "hono";

import { ShoppingAgent } from "../agent/agent";
import type { Env } from "../env";
import { loadSettings } from "../lib/config";
import { extractOtp, submitOtp } from "../lib/otp";
import type { AgentResponse, ChatRequest } from "../models/schemas";

const SSE_HEADERS = {
  "Content-Type": "text/event-stream",
  "Cache-Control": "no-cache",
  "X-Accel-Buffering": "no",
};

function sse(event: string, data: unknown): string {
  return `event: ${event}\ndata: ${JSON.stringify(data)}\n\n`;
}

function otpResponse(otp: string, threadId: string | null): AgentResponse {
  return {
    reply: `Submitted OTP **${otp}**.`,
    results: [],
    cart: [],
    checkout: null,
    pending_live: null,
    thread_id: threadId,
  };
}

export const agentRouter = new Hono<{ Bindings: Env }>();

agentRouter.post("/api/agent/chat", async (c) => {
  const body = await c.req.json<ChatRequest>();
  const settings = loadSettings(c.env);
  const threadId = body.thread_id ?? null;

  // OTP short-circuit — runs before (and instead of) the model.
  const latest = body.messages[body.messages.length - 1];
  if (latest?.role === "user") {
    const otp = extractOtp(latest.content);
    if (otp) {
      try {
        await submitOtp(settings, otp);
      } catch {
        return c.json({ detail: "Could not submit the OTP right now. Please try again." }, 502);
      }
      return c.json(otpResponse(otp, threadId));
    }
  }

  const agent = new ShoppingAgent(c.env, settings);
  try {
    return c.json(await agent.run(body.messages, threadId));
  } catch {
    return c.json({ detail: "The AI service is currently unavailable. Please try again." }, 502);
  }
});

agentRouter.post("/api/agent/chat/stream", async (c) => {
  const body = await c.req.json<ChatRequest>();
  const settings = loadSettings(c.env);
  const threadId = body.thread_id ?? null;
  const enc = new TextEncoder();
  const latest = body.messages[body.messages.length - 1];

  // OTP short-circuit (mirrors the non-streaming endpoint).
  if (latest?.role === "user") {
    const otp = extractOtp(latest.content);
    if (otp) {
      const stream = new ReadableStream({
        async start(controller) {
          try {
            await submitOtp(settings, otp);
            controller.enqueue(enc.encode(sse("done", otpResponse(otp, threadId))));
          } catch {
            controller.enqueue(
              enc.encode(sse("error", { detail: "Could not submit the OTP right now. Please try again." })),
            );
          }
          controller.close();
        },
      });
      return new Response(stream, { headers: SSE_HEADERS });
    }
  }

  const agent = new ShoppingAgent(c.env, settings);
  const stream = new ReadableStream({
    async start(controller) {
      try {
        for await (const [event, data] of agent.runStream(body.messages, threadId)) {
          controller.enqueue(enc.encode(sse(event, data)));
        }
      } catch {
        controller.enqueue(
          enc.encode(sse("error", { detail: "The AI service is currently unavailable. Please try again." })),
        );
      }
      controller.close();
    },
  });
  return new Response(stream, { headers: SSE_HEADERS });
});
